ext/nnstreamer/nnstreamer_protobuf/nnstreamer_protobuf.cc

```diff
@@ -0,0 +1,169 @@
+ * You may simply download binary packages from PPA
+ *
+ * $ sudo apt-add-repository ppa:nnstreamer
+ * $ sudo apt update
+ * $ sudo apt install libprotobuf-dev libprotobuf-lite17 libprotobuf17 protobuf-compiler17
+ */
+
+#include <nnstreamer_plugin_api.h>
+#include <nnstreamer_log.h>
+#include "nnstreamer.pb.h"      /* Generated by `protoc` */
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+namespace nnstreamer {
+namespace protobuf {
+
+/** @brief tensordec-plugin's GstTensorDecoderDef callback */
+GstFlowReturn
+gst_tensor_decoder_protobuf (const GstTensorsConfig * config,
+    const GstTensorMemory * input, GstBuffer * outbuf)
+{
+  GstMapInfo out_info;
+  GstMemory *out_mem;
+  size_t size, outbuf_size;
+
+  Tensors tensors;
+  Tensors::frame_rate *fr = NULL;
+
+  const unsigned int num_tensors = config->info.num_tensors;
+  g_assert (num_tensors > 0);
+  tensors.set_num_tensor (num_tensors);
+
+  fr = tensors.mutable_fr ();
+  if (!fr) {
+    nns_loge ("Failed to get pointer of tensors / tensordec-protobuf");
+    return GST_FLOW_ERROR;
+  }
+
+  fr->set_rate_n (config->rate_n);
+  fr->set_rate_d (config->rate_d);
+  
+  for (unsigned int i = 0; i < num_tensors; ++i) {
+    Tensor *tensor = tensors.add_tensor ();
+    gchar *name = NULL;
+
+    name = config->info.info[i].name;
+    if (name == NULL) {
+      tensor->set_name ("Anonymous");
+    } else {
+      tensor->set_name (name);
+    }
+
+    tensor->set_type ((Tensor::Tensor_type)
+                          config->info.info[i].type);
+
+    for (int j = 0; j < NNS_TENSOR_RANK_LIMIT; ++j) {
+      tensor->add_dimension (config->info.info[i].dimension[j]);
+    }
+
+    tensor->set_data (input[i].data, (int) input[i].size);
+  }
+
+  g_assert (outbuf);
+
+  size = tensors.ByteSizeLong ();
+  outbuf_size = gst_buffer_get_size (outbuf);
+  
+  if (outbuf_size == 0) {
+    out_mem = gst_allocator_alloc (NULL, size, NULL);
+  } else {
+    if (outbuf_size < size) {
+      gst_buffer_set_size (outbuf, size);
+    }
+    out_mem = gst_buffer_get_all_memory (outbuf);
+  }
+
+  if (FALSE == gst_memory_map (out_mem, &out_info, GST_MAP_WRITE)) {
+    nns_loge ("Cannot map output memory / tensordec-protobuf");
+    return GST_FLOW_ERROR;
+  }
+
+  tensors.SerializeToArray (out_info.data, size);
+
+  gst_memory_unmap (out_mem, &out_info);
+
+  if (gst_buffer_get_size (outbuf) == 0) {
+    gst_buffer_append_memory (outbuf, out_mem);
+  }
+  
+  return GST_FLOW_OK;
+}
+
+/** @brief tensor converter plugin's NNStreamerExternalConverter callback */
+GstBuffer *
+gst_tensor_converter_protobuf (GstBuffer * in_buf, gsize * frame_size,
+    guint * frames_in, GstTensorsConfig * config)
+{
+  Tensors tensors;
+  Tensors::frame_rate * fr = NULL;
+  GstMemory *in_mem, *out_mem;
+  GstMapInfo in_info;
+  GstBuffer *out_buf;
+  guint mem_size;
+  gpointer mem_data;
+
+  in_mem = gst_buffer_peek_memory (in_buf, 0);
+  g_assert (gst_memory_map (in_mem, &in_info, GST_MAP_READ));
```

Comment:
This is dangerous. Put out gst_memory_map from g_assert().

Reply:
Thank you! I modified this line and your other comments.